src/main/java/org/junitpioneer/jupiter/report/PioneerReportListener.java

```diff
@@ -0,0 +1,220 @@
+			// Store result in cache
+			testStatusCache.put(testIdentifier.getUniqueId(), testExecutionResult.getStatus().toString());
+		}
+	}
+
+	@Override
+	public void testPlanExecutionFinished(TestPlan testPlan) {
+		// Store plan to access it from methods
+		this.storedTestPlan = testPlan;
+		pioneerReport = new PioneerReport();
+
+		Set<TestIdentifier> allRoots = testPlan.getRoots();
+
+		for (TestIdentifier identifier : allRoots) {
+			if (identifier.isContainer()) {
+				pioneerReport.getTestcontainer().add(processContainer(identifier));
+			} else if (identifier.isTest()) {
+				pioneerReport.getTestcase().add(processTest(identifier));
+			}
+		}
+		writeReport(pioneerReport);
+	}
+
+	/**
+	 * Converts an {@code TestIdentifier} object into a {@code Testcontainer} for the report.
+	 *
+	 * @param container Object to be converted into {@code Testcontainer}
+	 * @return Created object
+	 */
+	Testcontainer processContainer(TestIdentifier container) {
+		Testcontainer testcontainer = new Testcontainer(container.getDisplayName());
+
+		// Add properties, if there are any
+		String uniqueId = container.getUniqueId();
+		Properties allProps = getProperties(uniqueId);
+
+		if (!allProps.getProperty().isEmpty()) {
+			testcontainer.properties = allProps;
+		}
+
+		Set<TestIdentifier> allChildren = storedTestPlan.getChildren(uniqueId);
+
+		for (TestIdentifier identifier : allChildren) {
+			if (identifier.isContainer()) {
+				testcontainer.getTestcontainer().add(processContainer(identifier));
+			} else if (identifier.isTest()) {
+				testcontainer.getTestcase().add(processTest(identifier));
+			}
+		}
+
+		return testcontainer;
+	}
+
+	/**
+	 * Converts an {@code TestIdentifier} object into a {@code Testcase} for the report.
+	 *
+	 * @param test Object to be converted into {@code Testcase} element
+	 * @return Created object
+	 */
+	Testcase processTest(TestIdentifier test) {
+		String status = testStatusCache.getOrDefault(test.getUniqueId(), "UNKNOWN");
+		Testcase testcase = new Testcase(test.getDisplayName(), status);
+
+		// Add properties, if there are any
+		Properties allProps = getProperties(test.getUniqueId());
+
+		if (!allProps.getProperty().isEmpty()) {
+			testcase.setProperties(allProps);
+		}
+
+		return testcase;
+	}
+
+	/**
+	 * Retrieves all cached properties for an {@code TestIdentifier} by its unique id.
+	 *
+	 * @param testIdentifierUniqueId Unique id the properties should be loaded for.
+	 * @return Wrapper which contains all cached properties or an empty list.
+	 */
+	Properties getProperties(String testIdentifierUniqueId) {
+		Properties allProperties = new Properties();
+
+		if (propertyCache.containsKey(testIdentifierUniqueId)) {
+			allProperties.getProperty().addAll(propertyCache.get(testIdentifierUniqueId));
+		}
+
+		return allProperties;
+	}
+
+	/**
+	 * Creates the report and validates it against the XSD.
+	 *
+	 * @param report Report to be created.
+	 */
+	void writeReport(PioneerReport report) {
```

Comment:
Shouldn't this method be `private`? (This is just copy-paste at this point. Feel free to address these concerns in a single comment.)

Reply:
I don't create private methods, so tests are able to access them.